app/uk/gov/hmrc/gform/testonly/TestOnlyFormService.scala

```diff
@@ -40,17 +51,64 @@ class TestOnlyFormService(
           val newId = java.util.UUID.randomUUID().toString
           cacheItem.data.fields.toList
             .traverse { case (k, v) =>
-              val value = if (k === "form") { v.as[JsObject] - "_id" - "userId" }
-              else v
-              snapshotRepository.put(newId)(DataKey(k), value)
+              if (k === "form") {
+                val formTemplateId = v.as[JsObject].value("formTemplateId").as[String]
+                val prefix = UniqueStringGenerator.generateUniqueString()
+                val updatedFormTemplateId = s"${prefix}_$formTemplateId"
+                val js = v.as[JsObject] - "_id" - "userId" ++ Json.obj("formTemplateId" -> updatedFormTemplateId)
+                for {
+                  _   <- snapshotRepository.put(newId)(DataKey(k), js)
+                  raw <- formTemplateService.get(FormTemplateRawId(formTemplateId))
+                  updatedRaw = raw.value ++ Json.obj("_id" -> updatedFormTemplateId)
+                  _ <- snapshotRepository.put(newId)(DataKey("template"), updatedRaw)
+                  _ <- restoreSnapshotTemplate(newId)
+                } yield ()
+              } else {
+                snapshotRepository.put(newId)(DataKey(k), v)
+              }
             }
-            .flatMap(_ => snapshotRepository.put(newId)(DataKey("description"), saveRequest.description))
+            .flatMap(_ =>
+              snapshotRepository.put(newId)(
+                DataKey("description"),
+                saveRequest.description
+              )
+            )
+            .flatMap(_ =>
+              snapshotRepository.put(newId)(
+                DataKey("gformVersion"),
+                Json.toJson(BuildInfo.version)
+              )
+            )
+            .flatMap(_ =>
+              snapshotRepository.put(newId)(
+                DataKey("gformFrontendVersion"),
+                saveRequest.gformFrontendVersion
+              )
+            )
             .map(_ => newId)
         case None =>
           Future.failed(new Exception(s"We could not find cache item with id: ${saveRequest.formId}"))
       }
       .flatMap(snapShotId => getSnapshotData(snapShotId))
 
+  def restoreSnapshotTemplate(snapshotId: String): Future[Unit] = {
+    val snapshotItem = snapshotRepository.findById(snapshotId)
```

Comment:
Why this returns some `CacheItem` which is not our type?

Can `snapshotRepository` operate on our type instead?

Reply:
We can, yes. But I thought it is unnecessary. We just manipulate jsons that represent `Form` or `Template` in the snapshot  and I thought to do that the functionality that bare mongo cache repository is adequate. We don't need to wrap/unwrap it to our data.

Reply:
There is at least 5 fields I can see at the moment. Let's use proper type for it.

Reply:
ok